### PR TITLE
Add loop contracts and harness for `small_slice_eq`

### PR DESCRIPTION
**Pull request by qinheping**
Porting from https://github.com/model-checking/kani/blob/main/tests/expected/loop-contract/small_slice_eq.rs

The cbmc argument `--object-bits 8` is needed to avoid running out of memory for the harness.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

**Comment by qinheping** [comment]
The method `small_slice_eq` and the harness only runs for configuration `target_arch = "x86_64"` and `target_feature = "sse2"`, and hence is not tested in the current CI checks.

File: library/core/src/str/pattern.rs [hunk 1]
```diff
@@ -1940,3 +1947,17 @@ unsafe fn small_slice_eq(x: &[u8], y: &[u8]) -> bool {
...
+#[unstable(feature = "kani", issue = "none")]
+pub mod verify {
+    #[cfg(all(target_arch = "x86_64", target_feature = "sse2"))] // only called on x86
+    #[kani::proof]
+    pub fn check_small_slice_eq() {
+        let _ = Box::new(0);
+        const ARR_SIZE: usize = 1000;
+        let x: [i32; ARR_SIZE] = kani::any();
+        let y: [i32; ARR_SIZE] = kani::any();
+        small_slice_eq(x, y);
```
Comment (celinval) [inline library/core/src/str/pattern.rs:1961]: Don't you need an unsafe block here?

File: library/core/src/str/pattern.rs [hunk 2]
```diff
@@ -1940,3 +1947,17 @@ unsafe fn small_slice_eq(x: &[u8], y: &[u8]) -> bool {
         vx == vy
     }
 }
+
+#[cfg(kani)]
+#[unstable(feature = "kani", issue = "none")]
+pub mod verify {
+    #[cfg(all(target_arch = "x86_64", target_feature = "sse2"))] // only called on x86
+    #[kani::proof]
+    pub fn check_small_slice_eq() {
```
Comment (celinval) [inline library/core/src/str/pattern.rs:1956]: It might be nice to add a bit of non-determinism to the slice sizes. You can maybe use `any_slice`.

**Review by celinval — COMMENTED** [review]
Can you also please add a contract to this function since it is unsafe?

File: library/core/src/str/pattern.rs (hunk 1 above)
Comment (qinheping) [reply library/core/src/str/pattern.rs]: Yes. I didn't find a good way to test it. Now I am running it by dropping the cfg constraint `target="sse2"`.

File: library/core/src/str/pattern.rs (hunk 2 above)
Comment (qinheping) [reply library/core/src/str/pattern.rs:1982]: Yes, it is a good idea. However, `any_slice_from_array` and `any_vec` are not in `kani_core`. So I copied their definitions from kani. Where is the better place to put them?

Comment (celinval) [reply library/core/src/str/pattern.rs:1982]: Move them to `kani_core`. :smile:

Comment (qinheping) [reply library/core/src/str/pattern.rs:1979]: Yes, I created a PR for it. https://github.com/model-checking/kani/pull/3646

File: library/core/src/lib.rs [hunk 3]
```diff
@@ -230,6 +230,7 @@
 #![feature(unboxed_closures)]
 #![feature(unsized_fn_params)]
 #![feature(with_negative_coherence)]
+#![feature(proc_macro_hygiene)]
```
Comment (celinval) [inline library/core/src/lib.rs:233]: why do we need this? I would prefer not enabling any new feature in the standard library

File: library/core/src/str/pattern.rs [hunk 4]
```diff
@@ -43,6 +43,14 @@ use crate::convert::TryInto as _;
 use crate::slice::memchr;
 use crate::{cmp, fmt};
 
+#[cfg(all(target_arch = "x86_64", any(kani, target_feature = "sse2")))] // only called on x86
+use safety::{loop_invariant, requires};
+
+#[cfg(kani)]
+use crate::kani;
+#[cfg(kani)]
+use crate::kani::mem::same_allocation;
```
Comment (celinval) [inline library/core/src/str/pattern.rs:52]: please use `ub_checks` instead.

File: library/core/src/str/pattern.rs [hunk 5]
```diff
@@ -1965,3 +1978,26 @@ unsafe fn small_slice_eq(x: &[u8], y: &[u8]) -> bool {
...
+
+#[cfg(kani)]
+#[unstable(feature = "kani", issue = "none")]
+pub mod verify {
+    use super::*;
+
+    #[cfg(all(kani, target_arch = "x86_64"))] // only called on x86
+    #[kani::proof]
+    #[kani::unwind(4)]
+    pub fn check_small_slice_eq() {
```
Comment (celinval) [inline library/core/src/str/pattern.rs:1990]: please add harnesses for empty slices with dangling pointer.

File: library/core/src/lib.rs (hunk 3 above)
Comment (qinheping) [reply library/core/src/lib.rs:233]: It is required to use macro for loops.

Comment (qinheping) [reply library/core/src/lib.rs:233]: Error I got without the feature suggests to include it.
```
error[E0658]: custom attributes cannot be applied to expressions
  --> simple_while_loop_not_enabled.rs:15:5
   |
15 |     #[kani::loop_invariant(x >= 2)]
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: see issue #54727 <https://github.com/rust-lang/rust/issues/54727> for more information
   = help: add `#![feature(proc_macro_hygiene)]` to the crate attributes to enable
   = note: this compiler was built on 2024-11-02; consider upgrading it if it is out of date
```

File: library/core/src/str/pattern.rs (hunk 5 above)
Comment (qinheping) [reply library/core/src/str/pattern.rs:1989]: I added one such test. But it fails and gives the same error even without loop contracts. I suppose another safety precondition of `small_slice_eq` should be not dangling pointer.

File: library/core/src/lib.rs (hunk 3 above)
Comment (celinval) [reply library/core/src/lib.rs:233]: Yes, in that case, I will suggest that we go ahead and merge this as is, and fix this in Kani later. Can you change this to conditionally add it only if kani is enabled? I.e.:
```suggestion
#![cfg_attr(kani, feature(proc_macro_hygiene))]
```

File: library/core/src/str/pattern.rs [hunk 6]
```diff
@@ -1965,3 +1977,45 @@ unsafe fn small_slice_eq(x: &[u8], y: &[u8]) -> bool {
...
+
+    #[cfg(all(kani, target_arch = "x86_64"))] // only called on x86
+    #[kani::proof]
+    #[kani::unwind(4)]
+    pub fn check_small_slice_eq_empty() {
+        let ptr_x = kani::any_where::<usize, _>(|val| *val != 0) as *const u8;
+        let ptr_y = kani::any_where::<usize, _>(|val| *val != 0) as *const u8;
+        kani::assume(ptr_x.is_aligned());
+        kani::assume(ptr_y.is_aligned());
+        unsafe {
```
Comment (celinval) [inline library/core/src/str/pattern.rs:2011]: please move the unsafe to be surrounding `small_slice_eq` call only.

**Review by celinval — APPROVED** [review]

File: library/core/src/lib.rs (hunk 3 above)
Comment (qinheping) [reply library/core/src/lib.rs:233]: It works for Kani proof but will fail the Rust Tests.

File: library/core/src/str/pattern.rs [hunk 7]
```diff
@@ -1965,3 +1977,45 @@ unsafe fn small_slice_eq(x: &[u8], y: &[u8]) -> bool {
...
 }
+
+#[cfg(kani)]
+#[unstable(feature = "kani", issue = "none")]
+pub mod verify {
+    use super::*;
+
+    #[cfg(all(kani, target_arch = "x86_64"))] // only called on x86
+    #[kani::proof]
+    #[kani::unwind(4)]
```
Comment (feliperodri) [inline library/core/src/str/pattern.rs:1988]: Why do we still need bounds in this case?

File: library/core/src/str/pattern.rs [hunk 8]
```diff
@@ -1965,3 +1977,45 @@ unsafe fn small_slice_eq(x: &[u8], y: &[u8]) -> bool {
...
+    pub fn check_small_slice_eq() {
+        // TODO: ARR_SIZE can be `std::usize::MAX` with cbmc argument
+        // `--arrays-uf-always`
```
Comment (feliperodri) [inline library/core/src/str/pattern.rs:1991]: Why not using it then?

File: library/core/src/str/pattern.rs [hunk 9]
```diff
@@ -1965,3 +1977,45 @@ unsafe fn small_slice_eq(x: &[u8], y: &[u8]) -> bool {
...
+    }
+
+    #[cfg(all(kani, target_arch = "x86_64"))] // only called on x86
+    #[kani::proof]
+    #[kani::unwind(4)]
+    pub fn check_small_slice_eq_empty() {
+        let ptr_x = kani::any_where::<usize, _>(|val| *val != 0) as *const u8;
+        let ptr_y = kani::any_where::<usize, _>(|val| *val != 0) as *const u8;
+        kani::assume(ptr_x.is_aligned());
+        kani::assume(ptr_y.is_aligned());
```
Comment (feliperodri) [inline library/core/src/str/pattern.rs:2014]: I feel this logic should also be integrated into `kani::slice::any_slice_of_array`. Any reason why not?

**Review by feliperodri — APPROVED** [review]

File: library/core/src/str/pattern.rs (hunk 8 above)
Comment (qinheping) [reply library/core/src/str/pattern.rs:1991]: We can't specify it for certain harness for now, and we don't want to use it for all harness.

File: library/core/src/str/pattern.rs (hunk 7 above)
Comment (qinheping) [reply library/core/src/str/pattern.rs:1988]: There is another loop in `small_slice_eq` for the case that the size of slice is less than 4.

File: library/core/src/str/pattern.rs (hunk 9 above)
Comment (qinheping) [reply library/core/src/str/pattern.rs:2014]: I think `any_slice_of_array` should be pointing to the array even when the size of the slice is zero. But I agree that we should have `any_slice` that can return empty slice of dangling pointers.

**Pull request merged**